orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -869,89 +869,216 @@ common::Status HandleSharedInitializer(Graph& graph,
+  std::string cut_index_str = std::to_string(current_rank);
+
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<bool>(graph,
+                                            "send_input_signal" + cut_index_str,
+                                            ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                                            true, /* initializer data */
+                                            send_input_args,
+                                            new_input_names));
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<size_t>(graph,
+                                              "send_dst_rank" + cut_index_str,
+                                              ONNX_NAMESPACE::TensorProto_DataType_INT64,
+                                              next_rank, /* initializer data */
+                                              send_input_args,
+                                              new_input_names));
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<bool>(graph,
+                                            "recv_input_signal" + cut_index_str,
+                                            ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                                            true, /* initializer data */
+                                            recv_input_args,
+                                            new_input_names));
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<size_t>(graph,
+                                              "recv_src_rank" + cut_index_str,
+                                              ONNX_NAMESPACE::TensorProto_DataType_INT64,
+                                              current_rank, /* initializer data */
+                                              recv_input_args,
+                                              new_input_names));
+
+  // add output node_arg for send/recv
+  AddNewNodeArg(graph, "send_output_signal" + cut_index_str,
+                ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                send_output_args, new_output_names);
+
+  AddNewNodeArg(graph, "receive_output_signal" + cut_index_str,
+                ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                recv_output_args, new_output_names);
+
+  return Status::OK();
+}
+
+common::Status SplitGraphWithMap(Graph& graph,
+                                 std::map<Node*, int>& op_to_rank,
+                                 int nstages,
+                                 std::vector<std::pair<int, int>>& messages,
+                                 std::vector<Node*>& send_nodes,
+                                 std::vector<Node*>& recv_nodes) {
+
+  // forward_messages[s]: all the tensors sent by stage s while executing
+  // forward computation.
+  std::vector<std::set<const NodeArg*>> forward_messages(nstages);
+  // TODO(jufranc): once we start using this function on the training graph,
+  // we need to keep backward_messages[s] too.
```

Comment:
We're using 'rank', 'stage', and 'device' in this function -- do they mean exactly the same thing?

Reply:
It should be always 'stage'. I'll try to rename all of them before opening the real PR.